### PR TITLE
Implement (--debug=json) JSON debug time output

### PR DESCRIPTION
**Pull request by bdbaddog**
Allow writing some of the `--debug` option outputs to a JSON file

`--debug=json,time` for example would write the build info and --debug=time info into a json file.

Added `DebugOptions(json='file/path')`

Done:
* [X] Not sure if we need to provide an API to change the file name/location?

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by bdbaddog** [comment]
Attaching example output file for `--debug=count,memory,time,action-timestamps,json` (basically enabling all parts of the file)

[scons_stats.json.txt](https://github.com/SCons/scons/files/11654969/scons_stats.json.txt)

File: SCons/Util/stats.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+import sys
+from datetime import datetime
+
+import SCons.Debug
+
+ALL_STATS = {}
+ENABLE_JSON = False
+JSON_OUTPUT_FILE = 'scons_stats.json'
+
+def AddStatType(name, stat_object):
```
Comment (mwichmann) [inline SCons/Util/stats.py]: A thought (don't have to follow): for new code, could pick names that would quiet pylint?  Know from experience that this one will generate:
```
Function name "AddStatType" doesn't conform to snake_case naming style (invalid-name)
```
Class names are fine as CamelCase, though.

File: SCons/Script/Main.py [hunk 2]
```diff
@@ -517,6 +519,26 @@ def GetOption(name):
...
+    if json is not None:
+        json_node = SCons.Defaults.DefaultEnvironment().arg2nodes(json)
+        SCons.Util.stats.JSON_OUTPUT_FILE = json_node[0].get_abspath()
+        # Check if parent dir to JSON_OUTPUT_FILE exists
+        json_dir = os.path.dirname(SCons.Util.stats.JSON_OUTPUT_FILE)
+        try:
+            if not os.path.isdir(json_dir):
+                os.makedirs(json_dir, exist_ok=True)
+            # Now try to open file and see if you can..
+            open(SCons.Util.stats.JSON_OUTPUT_FILE,'w')
```
Comment (mwichmann) [inline SCons/Script/Main.py]: Do you expect this to remain open from here? Or is this just a quick test-open? It's opened "for real" in `Util`, right?  Suggest using a `with` block here for clean resource allocation.

File: doc/man/scons.xml [hunk 3]
```diff
@@ -924,6 +924,21 @@ of a given derived file:</para>
...
+  </listitem>
+  </varlistentry>
+
+<varlistentry>
+  <term><emphasis role="bold">json</emphasis></term>
+  <listitem>
+<para>Write info to a JSON file for any of the following debug options if they are enabled: <emphasis>memory</emphasis>,
+    <emphasis>count</emphasis>, <emphasis>time</emphasis>, <emphasis>action-timestamps</emphasis> </para>
+    <para>The default output file is in the top level directory named <literal>scons_stats.json</literal></para>
+    <para>The file name/path can be modified by using <literal>DebugOptions(json='path/to/file.json')</literal></para>
```
Comment (mwichmann) [inline doc/man/scons.xml]: use link here. Suggest: "The name/path of the debug output file can be set using the `&f-link-DebugOptions;` function.

File: doc/man/scons.xml [hunk 4]
```diff
@@ -924,6 +924,21 @@ of a given derived file:</para>
...
 $ <userinput>scons --debug=includes foo.o</userinput>
+</screen>
+
+  </listitem>
+  </varlistentry>
+
+<varlistentry>
+  <term><emphasis role="bold">json</emphasis></term>
+  <listitem>
+<para>Write info to a JSON file for any of the following debug options if they are enabled: <emphasis>memory</emphasis>,
```
Comment (mwichmann) [inline doc/man/scons.xml]: Does the user need to be informed of the layout of the json file? In other words, if they want to examine the contents programmatically, do they get to deduce what will be in it, or should we tell them? (if the latter, not sure where it goes)

File: SCons/Util/stats.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                                  'duration': finish_time - start_time}
+
+
+COUNT_STATS = CountStats()
+MEMORY_STATS = MemStats()
+TIME_STATS = TimeStats()
+
+
+
+def WriteJsonFile():
```
Comment (mwichmann) [inline SCons/Util/stats.py]: nitpickery, but "Write JSON file" is pretty context-specific (we wrte other json files - in msvc for example). I'd prefer to see the word "debug" or "stats" in there somewhere. Even though this is activated by debug flags, it's really stats, so maybe `write_stats_file`?

File: doc/man/scons.xml [hunk 6]
```diff
@@ -924,6 +924,21 @@ of a given derived file:</para>
...
+    <emphasis>count</emphasis>, <emphasis>time</emphasis>, <emphasis>action-timestamps</emphasis> </para>
+    <para>The default output file is in the top level directory named <literal>scons_stats.json</literal></para>
```
Comment (mwichmann) [inline doc/man/scons.xml]: Feels a little awkward?  Maybe swap chunks?: "The default output file is `<literal>scons_stats.json</literal>` in the top-level directory."

File: SCons/Util/stats.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    """
+    Add a statistic type to the global collection
+    """
+    if name in ALL_STATS:
+        raise UserWarning(f'Stat type {name} already exists')
+    else:
+        ALL_STATS[name] = stat_object
+
+
+class Stats:
```
Comment (mwichmann) [inline SCons/Util/stats.py]: Might as well explicitly make it an ABC since you're using it that way...

File: SCons/Util/stats.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+
+
+def WriteJsonFile():
+    """
+    Actually write the JSON file with debug information.
+    Depending which of : count, time, action-timestamps,memory their information will be written.
+    """
+
+    from SCons.Script import BUILD_TARGETS, COMMAND_LINE_TARGETS, ARGUMENTS, ARGLIST
```
Comment (mwichmann) [inline SCons/Util/stats.py]: perhaps annotate to avoid checker `import-outside-toplevel` type errors?

File: SCons/Util/stats.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+    def add_command(self, command, start_time, finish_time):
+        if command in self.commands:
+            print("Duplicate command %s" % command)
+        self.commands[command] = {'start': start_time,
+                                  'end' : finish_time,
+                                  'duration': finish_time - start_time}
+
+
+COUNT_STATS = CountStats()
```
Comment (mwichmann) [inline SCons/Util/stats.py]: A little fuzzy on Python naming conventions... yes, PEP8 says to use CAPS for module-level constants, but these aren't really constants - they're updated elsewhere during use.  I'll just leave this here without a recommendation (sorry!)....

**Review by mwichmann — COMMENTED** [review]
This seems to look fine in general.  I know there was some stuff in discord about maybe supporting other formats (I hope, in 2023, that we don't need to output xml!!!)

File: SCons/Util/stats.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    # Now add information about this build to the JSON file
+    json_structure['Build_Info'] = {
+        'BUILD_TARGETS' : [str(t) for t in BUILD_TARGETS],
+        'ARGUMENTS' : [str(a) for a in ARGUMENTS],
+        'ARGLIST' : [ str(al) for al in ARGLIST],
+        'COMMAND_LINE_TARGETS' : [ str(clt) for clt in COMMAND_LINE_TARGETS],
+        'ARGV' : sys.argv,
+        'TIME' : datetime.now().isoformat(),
+        'HOST' : platform.node(),
+        'PYTHON_VERSION' : {
```
Comment (mwichmann) [inline SCons/Util/stats.py:197]: I'd use `sysconfig` to pull the info (more "standardized" that way) and include more details, although I don't specifically know which ones to add. Would people use this to compare behavior of different Python builds?

File: SCons/Script/Main.py (hunk 2 above)
Comment (bdbaddog) [reply SCons/Script/Main.py]: yes. good point. Was just trying to cover all bases on whether the specified file could be written to.

File: doc/man/scons.xml (hunk 3 above)
Comment (bdbaddog) [reply doc/man/scons.xml]: Added link and then prefixed example with "for example"

File: doc/man/scons.xml (hunk 4 above)
Comment (bdbaddog) [reply doc/man/scons.xml]: I think best left as exercise for the reader?.. manpage is not the right place as it's fairly lengthy..

File: doc/man/scons.xml (hunk 6 above)
Comment (bdbaddog) [reply doc/man/scons.xml]: agreed. changed.

File: SCons/Util/stats.py (hunk 7 above)
Comment (bdbaddog) [reply SCons/Util/stats.py]: Made change. Lemme know if that's the correct way to do?

File: SCons/Util/stats.py (hunk 9 above)
Comment (bdbaddog) [reply SCons/Util/stats.py]: good point. changing.

File: SCons/Util/stats.py (hunk 10 above)
Comment (bdbaddog) [reply SCons/Util/stats.py:197]: potentially you're storing these for a farm of build machines.. this way you could identify one being slower/faster and investigate.. at least that's what I was thinking.
Seemed simple enough to add it now, and could be useful..

File: SCons/Util/stats.py (hunk 8 above)
Comment (bdbaddog) [reply SCons/Util/stats.py]: done.

File: SCons/Util/stats.py (hunk 5 above)
Comment (bdbaddog) [reply SCons/Util/stats.py]: changed.

File: SCons/Util/stats.py (hunk 1 above)
Comment (bdbaddog) [reply SCons/Util/stats.py]: changed.

File: SCons/Script/Main.py (hunk 2 above)
Comment (mwichmann) [reply SCons/Script/Main.py]: yeah, we have other "test opens" - dblite does that for example.

File: SCons/Util/stats.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+class Stats(ABC):
+    def __init__(self):
+        self.stats = []
+        self.labels = []
+        self.append = self.do_nothing
+        self.print_stats = self.do_nothing
+        self.enabled = False
+
+    def do_append(self, label):
```
Comment (mwichmann) [inline SCons/Util/stats.py:67]: you can leave the `@abstractmethod` on these - that was in the original, no? (a little hard to find in GH web interface)

File: SCons/Util/stats.py (hunk 10 above)
Comment (mwichmann) [reply SCons/Util/stats.py:197]: sure, I was just suggesting more info might be good - or maybe add it later. For fun, do `python -m sysconfig` ... except on Windows, there's a TON of info captured now.

File: SCons/Util/stats.py (hunk 9 above)
Comment (mwichmann) [reply SCons/Util/stats.py]: changes what? I can't figure out if github isn't showing me the latest rev, but they still look in caps to me so no sure what changed.

File: SCons/Util/stats.py (hunk 10 above)
Comment (bdbaddog) [reply SCons/Util/stats.py:197]: Ugh.. that's too much info... I think I'll leave it as is for now.. ;)

File: SCons/Util/stats.py (hunk 11 above)
Comment (bdbaddog) [reply SCons/Util/stats.py:67]: nope. adding it breaks things.. messsaged in discord..

**Pull request merged**